Sources/OpenCombine/Publishers/Publishers.Decode.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  Publishers.Decode.swift
+//
+//
+//  Created by Joseph Spadafora on 6/21/19.
+//
+
+extension Publishers {
+
+    public struct Decode<Upstream, Output, Coder> : Publisher where
+        Upstream : Publisher,
+        Output : Decodable,
+        Coder : TopLevelDecoder,
+        Upstream.Output == Coder.Input {
+
+        /// The kind of errors this publisher might publish.
+        ///
+        /// Use `Never` if this `Publisher` does not publish errors.
+        public typealias Failure = Error
+
+        public let upstream: Upstream
+
+        internal let decoder: Coder
+
+        public init(upstream: Upstream, decoder: Coder) {
+            self.upstream = upstream
+            self.decoder = decoder
+        }
+
+        /// This function is called to attach the specified `Subscriber`
+        /// to this `Publisher` by `subscribe(_:)`
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<Receiver: Subscriber>(subscriber: Receiver)
+            where Failure == Receiver.Failure, Output == Receiver.Input {
+                let decodeSubscriber = _Decode<Upstream, Receiver, Coder>(
+                    downstream: subscriber,
+                    decoder: decoder
+                )
+                upstream.receive(subscriber: decodeSubscriber)
+        }
+    }
+}
+
+// swiftlint:disable:next line_length
+private final class _Decode<Upstream: Publisher, Downstream: Subscriber, Coder: TopLevelDecoder>:
+    OperatorSubscription<Downstream>, Subscriber,
+    CustomStringConvertible, Subscription
+    where
+    Downstream.Input: Decodable,
+    Coder.Input == Upstream.Output,
+    Downstream.Failure == Error {
+
+    typealias Input = Upstream.Output
+    typealias Failure = Upstream.Failure
+    typealias Output = Downstream.Input
+
+    private let _decoder: Coder
+    private var _demand: Subscribers.Demand = .none
```

Comment:
This property is not needed, as `Encode`/`Decode` have map-like semantics — they just republish their input by applying some transformation.

In fact, both operators can be implemented in terms of `TryMap`:

```swift
Publishers
    .Just(encoded)
    .decode(type: [Int].self, decoder: JSONDecoder())
// ↑ are equivalent ↓
Publishers
    .Just(encoded)
    .tryMap { try JSONDecoder().decode([Int].self, from: $0) }
```

If you look at the implementation of `Map`, you'll see that we don't change the requested demand in any way, we just re-request if from the upstream.

Reply:
Okay, the semantics make sense to me, but I'm not totally clear on how you'd like this to be changed to work.  Can you explain?

Reply:
Like in the `Map` implementation:

https://github.com/broadwaylamb/OpenCombine/blob/73071a2a9b507239345e781bd041ddd4abe96054/Sources/OpenCombine/Publishers/Publishers.Map.swift#L67-L70

https://github.com/broadwaylamb/OpenCombine/blob/73071a2a9b507239345e781bd041ddd4abe96054/Sources/OpenCombine/Publishers/Publishers.Map.swift#L80-L82

I. e. we just forward the method calls, hence no need to store the demand in this property.